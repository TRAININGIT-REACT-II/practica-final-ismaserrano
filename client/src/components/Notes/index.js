export { default as Notes } from "./Notes";
export { default as NotesAdd } from "./NotesAdd";
export { default as NotesDetail } from "./NotesDetail";
