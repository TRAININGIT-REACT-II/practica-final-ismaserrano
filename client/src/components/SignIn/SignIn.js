import Avatar from "@material-ui/core/Avatar";
import Button from "@material-ui/core/Button";
import CssBaseline from "@material-ui/core/CssBaseline";
import TextField from "@material-ui/core/TextField";
import Link from "@material-ui/core/Link";
import Grid from "@material-ui/core/Grid";
import LockOutlinedIcon from "@material-ui/icons/LockOutlined";
import Typography from "@material-ui/core/Typography";
import Container from "@material-ui/core/Container";
import useStyles from "./styles";
import useApi from "../../hooks/useApi";
import { useForm, Controller } from "react-hook-form";
import { useHistory } from "react-router-dom";
import { useDispatch } from "react-redux";

/* Acciones */
import { setUser } from "../../actions/user";
import { useEffect, useState } from "react";

const SignIn = () => {
  const classes = useStyles();
  const loginReq = useApi("/api/login", "", {}, false);
  const {
    handleSubmit,
    control,
    formState: { errors },
  } = useForm();
  const history = useHistory();
  const dispatch = useDispatch();
  const [apiError, setApiError] = useState("");

  let token;
  if (loginReq.data) {
    token = loginReq.data.token;
    if (token.length > 0) {
      sessionStorage.setItem("userObj", JSON.stringify(loginReq.data));
      dispatch(setUser(loginReq.data));
      history.push("/");
    }
  }

  const login = (data) => {
    loginReq.updateParams({
      method: "POST",
      body: JSON.stringify({
        username: data.email,
        password: data.password,
      }),
    });
    loginReq.perform();
  };

  const onSubmit = (data) => {
    login(data);
  };

  useEffect(() => {
    console.log(loginReq);
    if (
      !loginReq.loading &&
      loginReq.error !== null &&
      loginReq.error !== "" &&
      loginReq.data === null
    ) {
      setApiError(loginReq.error);
    }
  }, [loginReq]);

  return (
    <Container component="main" maxWidth="xs">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h5">
          Acceder
        </Typography>
        <form
          className={classes.form}
          noValidate
          onSubmit={handleSubmit(onSubmit)}
        >
          <Grid container spacing={2}>
            <Grid item xs={12}>
              <Controller
                name="email"
                render={({ field }) => (
                  <TextField
                    {...field}
                    variant="outlined"
                    fullWidth
                    id="email"
                    label="Email"
                    error={errors.email ? true : false}
                    helperText={errors.email ? errors.email.message : null}
                  />
                )}
                control={control}
                rules={{
                  required: {
                    value: true,
                    message: "Correo electrónico obligatorio",
                  },
                  pattern: {
                    value:
                      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
                    message: "Correo electrónico inválido",
                  },
                }}
              />
            </Grid>
            <Grid item xs={12}>
              <Controller
                name="password"
                render={({ field }) => (
                  <TextField
                    {...field}
                    variant="outlined"
                    required
                    fullWidth
                    label="Contraseña"
                    type="password"
                    id="password"
                    error={errors.password ? true : false}
                    helperText={
                      errors.password ? errors.password.message : null
                    }
                  />
                )}
                control={control}
                rules={{
                  required: {
                    value: true,
                    message: "La contaseña es obligatoria",
                  },
                  minLength: {
                    value: 8,
                    message:
                      "La contraseña debe tener una longitud mínima de 8 caracteres",
                  },
                  maxLength: {
                    value: 20,
                    message:
                      "La contraseña debe tener una longitud máxima de 20 caracteres",
                  },
                }}
              />
            </Grid>
          </Grid>
          <Button
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
          >
            Acceder
          </Button>
          {apiError.length > 0 ? (
            <Typography
              component="p"
              variant="body2"
              color="error"
              align="center"
            >
              {apiError}
            </Typography>
          ) : null}
          <Grid container>
            <Grid item>
              <Link href="/signup" variant="body2">
                Regístrate
              </Link>
            </Grid>
          </Grid>
        </form>
      </div>
    </Container>
  );
};

export default SignIn;
