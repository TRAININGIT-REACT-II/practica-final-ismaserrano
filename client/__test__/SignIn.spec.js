import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";

import { SignIn } from "../src/components/SignIn";

import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

const mockStore = configureStore([]);

describe("Componente SignIn", () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      myState: "sample text",
    });
  });

  it("Ha cambiado el renderizado de componente?", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <SignIn />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
