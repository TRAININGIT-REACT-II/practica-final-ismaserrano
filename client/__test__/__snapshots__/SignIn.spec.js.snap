// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente SignIn Ha cambiado el renderizado de componente? 1`] = `
<DocumentFragment>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs"
  >
    <div
      class="makeStyles-paper-1"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-2 MuiAvatar-colorDefault"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
          />
        </svg>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h5"
      >
        Acceder
      </h1>
      <form
        class="makeStyles-form-3"
        novalidate=""
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink="false"
                for="email"
                id="email-label"
              >
                Email
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="email"
                  name="email"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-7"
                  >
                    <span>
                      Email
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Contraseña
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-7"
                  >
                    <span>
                      Contraseña *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-4 MuiButton-containedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Acceder
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-body2 MuiTypography-colorPrimary"
              href="/signup"
            >
              Regístrate
            </a>
          </div>
        </div>
      </form>
    </div>
  </main>
</DocumentFragment>
`;
